File: attractor/pipeline/goal_gate.py
"""Goal gate enforcement."""

from __future__ import annotations

from pydantic import BaseModel, Field

from .graph import Graph, NodeType


class GoalGateResult(BaseModel):
    satisfied: bool = True
    unsatisfied_gates: list[str] = Field(default_factory=list)
    retry_target: str = ""


def check_goal_gates(
    graph: Graph,
    completed_outcomes: dict[str, str],
) -> GoalGateResult:
    """Check if all goal gate nodes achieved SUCCESS or PARTIAL_SUCCESS.

    Args:
        graph: The pipeline graph.
        completed_outcomes: Map of node_id -> outcome status string.
    """
    unsatisfied: list[str] = []
    acceptable = {"success", "partial_success"}

    for node in graph.nodes.values():
        if not node.goal_gate:
            continue
        outcome = completed_outcomes.get(node.id, "")
        if outcome not in acceptable:
            unsatisfied.append(node.id)

    if not unsatisfied:
        return GoalGateResult(satisfied=True)

    # Find retry target from the chain
    retry_target = ""
    for gate_id in unsatisfied:
        node = graph.get_node(gate_id)
        if node and node.retry_target:
            retry_target = node.retry_target
            break

    if not retry_target and graph.retry_target:
        retry_target = graph.retry_target
    if not retry_target and graph.fallback_retry_target:
        retry_target = graph.fallback_retry_target

    return GoalGateResult(
        satisfied=False,
        unsatisfied_gates=unsatisfied,
        retry_target=retry_target,
    )

File: internal/pipeline/goalgate.go
package pipeline

// GoalGateResult reports whether all goal gates are satisfied, and where to
// route otherwise.
type GoalGateResult struct {
	Satisfied        bool
	UnsatisfiedGates []string
	RetryTarget      string
}

// CheckGoalGates verifies every goal-gate node reached success or
// partial_success. completedOutcomes maps node id to outcome status.
func CheckGoalGates(g *Graph, completedOutcomes map[string]string) GoalGateResult {
	acceptable := map[string]bool{"success": true, "partial_success": true}
	var unsatisfied []string
	for _, id := range g.sortedNodeIDs() {
		node := g.Nodes[id]
		if !node.GoalGate() {
			continue
		}
		if !acceptable[completedOutcomes[node.ID]] {
			unsatisfied = append(unsatisfied, node.ID)
		}
	}
	if len(unsatisfied) == 0 {
		return GoalGateResult{Satisfied: true}
	}

	retryTarget := ""
	for _, gateID := range unsatisfied {
		if node := g.GetNode(gateID); node != nil && node.RetryTarget() != "" {
			retryTarget = node.RetryTarget()
			break
		}
	}
	if retryTarget == "" {
		retryTarget = g.RetryTarget
	}
	if retryTarget == "" {
		retryTarget = g.FallbackRetryTarget
	}

	return GoalGateResult{Satisfied: false, UnsatisfiedGates: unsatisfied, RetryTarget: retryTarget}
}
